Support for reviewers & assignees

StefMa opened the pull request:
fixes #215 

This time the forgotten reviewers and assignees 🙂 

Haven't tested it yet ...

alextu commented on src/main/java/org/gradle/wrapperupgrade/UpgradeWrapper.java:
```diff
@@ -201,6 +203,36 @@ private void gitCreatePr(Params params, String prTitle, String prBody) throws IO
+            if (!reviewers.isEmpty()) {
+                List<GHUser> githubReviewers = reviewers.stream()
+                    .map(reviewer -> {
+                        try {
+                            return params.gitHub.getUser(reviewer);
+                        } catch (IOException e) {
+                            getLogger().warn(String.format("Error fetching GitHub user '%s'", reviewer), e);
+                            return null;
+                        }
```
Suggesting to extract as a function and re-use for the assignees

StefMa replied:
Done here https://github.com/gradle/wrapper-upgrade-gradle-plugin/pull/271/commits/d4cbfb935661209acf1961d03a742b1348e10798

alextu reviewed (approved):
Thanks for your contribution! LGTM, just suggesting a small refactoring and we can merge it once you confirm it works properly (I'll test it too)

alextu commented on src/main/java/org/gradle/wrapperupgrade/UpgradeWrapper.java:
```diff
@@ -201,6 +203,14 @@ private void gitCreatePr(Params params, String prTitle, String prBody) throws IO
             }
+            List<String> reviewers = upgrade.getOptions().getReviewers().get();
+            if (!reviewers.isEmpty()) {
+                List<GHUser> githubReviewers = mapToGHUsers(params, reviewers);
```
I think you forgot to apply the reviewers to the PR with `pr.requestReviewers(githubReviewers)`, same for the assignees

StefMa commented:
> once you confirm it works properly 

I can confirm that "it works". 

Assignees works perfectly. 
See https://github.com/StefMa/wrapper-upgrade-gradle-plugin/pull/6

I “couldn’t test” reviewers. First, I tried to put you as a reviewer, then I got the message from GitHub (via the GitHub API project) that you can only add collaborators as reviewers. Then I set myself as a reviewer, but that didn’t work either. The message was that you can’t set the reviewer to the same person as the author.

Because those errors indicate that there are API calls happening, I would assume that everything works as expected.

However, those errors throw an exception and therefore failed the build. Maybe we want to catch those and report them only via the logger? 🤔 What do you think?

alextu commented:
> > once you confirm it works properly
> 
> I can confirm that "it works".
> 
> Assignees works perfectly. See [StefMa#6](https://github.com/StefMa/wrapper-upgrade-gradle-plugin/pull/6)
> 
> I “couldn’t test” reviewers. First, I tried to put you as a reviewer, then I got the message from GitHub (via the GitHub API project) that you can only add collaborators as reviewers. Then I set myself as a reviewer, but that didn’t work either. The message was that you can’t set the reviewer to the same person as the author.
> 
> Because those errors indicate that there are API calls happening, I would assume that everything works as expected.
> 
> However, those errors throw an exception and therefore failed the build. Maybe we want to catch those and report them only via the logger? 🤔 What do you think?

I've tested it as well and noticed the same, I'll be able to test it end to end tomorrow. I agree that we should not fail the build since the PR is created, and just output a warning message in that case.

StefMa commented:
I changed the code a bit and cleaned it up.
Check it out here: https://github.com/gradle/wrapper-upgrade-gradle-plugin/pull/271/commits/a2a408eb2350564d90a330e525aecc8f552b9b80

* Add try-catch around `addAssignees` and `requestReviewers` and `addLabels`
* In case of an error, we just log it as warning
* Also I moved the parts after the output of "PR created", because actually this happens afterwards
* I extract methods for each of them 

I tested the case of "request review for myself (author)" and the build succeeded, but the error is visible:
![Screenshot 2024-11-29 at 8 29 03 AM](https://github.com/user-attachments/assets/93f60cd7-bccd-4312-96c6-50827d33a6a3)

I tested to add a label that **doesn't exist** to check if an error happen.
But actually it just creates that label 🫠 
See https://github.com/StefMa/wrapper-upgrade-gradle-plugin/pull/8 🤷 
Anyways, according to the [method signature](https://github.com/hub4j/github-api/blob/107c8e6efe5c9d350bf6632302bce633bf548291/src/main/java/org/kohsuke/github/GHIssue.java#L408) it could also throw an `IOException`.
So better safe than sorry I would say.

alextu commented:
Thanks! I've tested it with reviewers as well, looks good, I'll merge the PR soon
